bitcoin/src/blockdata/transaction.rs

```diff
@@ -1135,6 +1157,65 @@ mod sealed {
     impl Sealed for super::Version {}
 }
 
+#[cfg(feature = "arbitrary")]
+impl<'a> Arbitrary<'a> for InputWeightPrediction {
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        let witness_element_lengths = Vec::<usize>::arbitrary(u)?;
+        let input_script_len:usize = u.int_in_range(0..=5)?;
+
+        // --- This is a checked version of InputWeightPrdiction::new() ---
+        let mut result = None;
+        if let Some((count, total_size)) = witness_element_lengths.clone().into_iter().try_fold(
+            (0usize, 0usize),
+            |(count, total_size), elem_len| {
+                let elem_len = *elem_len.borrow();
+                let elem_size_opt = elem_len.checked_add(compact_size::encoded_size(elem_len));
+                if let Some(elem_size) = elem_size_opt {
+                    if let Some(new_size) = total_size.checked_add(elem_size) {
+                        return Some((count + 1, new_size))
+                    }
+                }
+                None
+            },) {
+            let witness_size_opt =
+                if count > 0 {
+                    total_size.checked_add(compact_size::encoded_size(count))
+                } else {
+                    Some(0)
+                };
+            let script_size_opt = input_script_len.checked_add(compact_size::encoded_size(input_script_len));
+            if let Some(witness_size) = witness_size_opt {
+                if let Some(script_size) = script_size_opt {
+                    result = Some(InputWeightPrediction { script_size, witness_size });
+                }
+            }
+        }
+        // ---
+
+        if result.is_some() {
+            match u.int_in_range(0..=5)? {
+                0 => Ok(InputWeightPrediction::P2WPKH_MAX),
+                1 => Ok(InputWeightPrediction::NESTED_P2WPKH_MAX),
+                2 => Ok(InputWeightPrediction::P2PKH_COMPRESSED_MAX),
+                3 => Ok(InputWeightPrediction::P2PKH_UNCOMPRESSED_MAX),
+                4 => Ok(InputWeightPrediction::P2TR_KEY_DEFAULT_SIGHASH),
+                5 => Ok(InputWeightPrediction::P2TR_KEY_NON_DEFAULT_SIGHASH),
+                6 => Ok(InputWeightPrediction::new(input_script_len, witness_element_lengths)),
+                _ => Ok(InputWeightPrediction::from_slice(input_script_len, &witness_element_lengths)),
+            }
+        } else {
+            match u.int_in_range(0..=5)? {
+                0 => Ok(InputWeightPrediction::P2WPKH_MAX),
+                1 => Ok(InputWeightPrediction::NESTED_P2WPKH_MAX),
+                2 => Ok(InputWeightPrediction::P2PKH_COMPRESSED_MAX),
+                3 => Ok(InputWeightPrediction::P2PKH_UNCOMPRESSED_MAX),
+                4 => Ok(InputWeightPrediction::P2TR_KEY_DEFAULT_SIGHASH),
+                _ => Ok(InputWeightPrediction::P2TR_KEY_NON_DEFAULT_SIGHASH),
+            }
+        }
+    }
+}
```

Comment:
This all seems like a lot of code, why not just limit the values before calling `new` and `from_slice` so they don't panic?

Reply:
> This all seems like a lot of code, why not just limit the values before calling new and from_slice so they don't panic?

@tcharding I updated this so that it "limits the values".  It's still a lot of code and I was hoping to avoid writing code like this although I think this most closely matches what you and @apoelstra requested.

Reply:
actually I should probably make use of `take_while()`  to clean this up a bit.